tests/db/test_unit_db_models.py

```diff
@@ -0,0 +1,899 @@
+    for i, task in enumerate(db_workflow.task_list):
+        assert task.order == i
+
+    task4 = Task(**tasks_common_args, source="source4")
+    db.add(task4)
+    await db.commit()
+    await db_workflow.insert_task(
+        db=db, task_id=task4.id, order=1, meta={"meta": "test"}
+    )
+    db.expunge_all()
+
+    workflow_query = await db.execute(select(Workflow))
+    db_workflow = workflow_query.scalars().one()
+    assert len(db_workflow.task_list) == 4
+    assert db_workflow.task_list[1].meta == {"meta": "test"}
+
+    # test cascade (Workflow deletion removes all related WorkflowTasks)
+    await db.delete(db_workflow)
+    workflowtask_query = await db.execute(select(WorkflowTask))
+    db_workflowtask = workflowtask_query.scalars().one_or_none()
+    assert db_workflowtask is None
+
+
+async def test_project_and_datasets(db):
+
+    project = Project(name="project")
+    # using `.project` relationship
+    dataset1 = Dataset(name="dataset1", project=project)
+    db.add(project)
+    db.add(dataset1)
+    await db.commit()
+    db.expunge_all()
+
+    with pytest.raises(IntegrityError):
+        # missing relatioship with project
+        broken_dataset = Dataset(name="broken")
+        db.add(broken_dataset)
+        await db.commit()
+    await db.rollback()
+
+    project_query = await db.execute(select(Project))
+    db_project = project_query.scalars().one()
+    dataset_query = await db.execute(select(Dataset))
+    db_dataset1 = dataset_query.scalars().one()
+
+    # test relationships
+    assert db_dataset1.project_id == db_project.id
+    assert db_dataset1.project == db_project
+    # test defaults
+    assert db_dataset1.type is None
+    assert db_dataset1.read_only is False
+    assert db_dataset1.list_jobs_input == []
+    assert db_dataset1.list_jobs_output == []
+    assert db_dataset1.resource_list == []
+    assert db_dataset1.meta == {}
+    assert db_dataset1.history == []
+
+    # using `.project_id` attribute
+    dataset2 = Dataset(name="dataset2", project_id=db_project.id)
+    db.add(dataset2)
+    await db.commit()
+    db.expunge_all()
+
+    project_query = await db.execute(select(Project))
+    db_project = project_query.scalars().one()
+
+    dataset_query = await db.execute(select(Dataset))
+    db_dataset1, db_dataset2 = dataset_query.scalars().all()
+
+    # test relationships
+    assert db_dataset1.name == dataset1.name
+    assert db_dataset2.name == dataset2.name
+    assert db_dataset2.project_id == db_project.id
+    assert db_dataset2.project == db_project
+
+    # delete just one dataset
+    await db.delete(db_dataset2)
+
+    dataset_query = await db.execute(select(Dataset))
+    db_dataset = dataset_query.scalars().one()
+    assert db_dataset.name == dataset1.name
+
+    # delete the project
+    project_query = await db.execute(select(Project))
+    db_project = project_query.scalars().one()
+    await db.delete(db_project)
+
+    DB_ENGINE = Inject(get_settings).DB_ENGINE
```

Comment:
I was hesitant at first, but I now fully agree that this is a neat way of showing how things work. Thanks!